### ISSUE TITLE
Socket message processing slow.

### ISSUE THREAD

**Issue by UtkuX** [JKorf/Binance.Net#1035]
"2022/03/18 05:56:11:812 | Warning | Binance     | Socket 7 message processing slow (18250ms), consider offloading data handling to another thread. Data from this socket may arrive late or not at all if message processing is continuously slow."

Hello, while trying to get data over socket using this wrapper, it gave the above warning in the debug output section. However, I was already doing the work of pulling data from the socket with a separate task. do you have any solution suggestions? thanks...

**Comment by JKorf** [comment]
I'm not sure how you've implemented it, but the issue is that the action inside the data handler of the socket subscription is taking 18 seconds to complete. During these 18 seconds the message processing thread can't process any other messages, so they might not arrive in your application on time. 

You should try to make the data handling part of the socket subscription as short as possible. For example, if you want to execute some code that might take a little while when an order update comes in, you should do something like this: (pseudocode)
````csharp
var processQueue = new ConcurrentQueue<Order>();
socketClient.SpotStreams.SubscribeToUserDataUpdates(key, data =>{
  processQueue.Enqueue(data.Data);
});
````
and have another thread process the items in the queue. Hope that clears it up.

**Comment by UtkuX** [comment]
actually quite a novice. I created a class like this, then I want to use this data in program.cs. Is there any wrong place?


```
public class BinanceCoins
{
    public Dictionary<string, Tuple<decimal, decimal, decimal, decimal>> KlineDictionary = new Dictionary<string, Tuple<decimal, decimal, decimal, decimal>>();
    public Dictionary<string, Tuple<string, decimal>> TickerDictionary = new Dictionary<string, Tuple<string, decimal>>();

    public BinanceCoins(List<string> symbol)
    { 
        {
            BinanceSocketClient socket = new BinanceSocketClient();
            _ = Task.Run(() =>
              {
                  _ = socket.Spot.SubscribeToKlineUpdatesAsync(symbol, KlineInterval.OneMinute, data =>
                  {
                      if (KlineDictionary.ContainsKey(data.Data.Symbol))
                      {
                          KlineDictionary[data.Data.Symbol] = new Tuple<decimal, decimal, decimal, decimal>((data.Data.Data.High-data.Data.Data.Open)/data.Data.Data.Open,(data.Data.Data.High-data.Data.Data.Low)/data.Data.Data.Low ,(data.Data.Data.Close - data.Data.Data.Open) / data.Data.Data.Open, data.Timestamp.Second);
                      }
                      else
                      {
                          KlineDictionary.Add(data.Data.Symbol, new Tuple<decimal, decimal, decimal, decimal>((data.Data.Data.High-data.Data.Data.Open)/data.Data.Data.Open,(data.Data.Data.High-data.Data.Data.Low)/ data.Data.Data.Low ,(data.Data.Data.Close - data.Data.Data.Open) / data.Data.Data.Open, data.Timestamp.Second));
                      }
                  });
              });

            _ = Task.Run(() =>
              {
                  _ = socket.Spot.SubscribeToAggregatedTradeUpdatesAsync(symbol, data =>
                  {
                      if (TickerDictionary.ContainsKey(data.Data.Symbol))
                      {
                          TickerDictionary[data.Data.Symbol] = new Tuple<string, decimal>(data.Data.Symbol, data.Data.Price);
                      }
                      else
                      {
                          TickerDictionary.Add(data.Data.Symbol, new Tuple<string, decimal>(data.Data.Symbol, data.Data.Price));
                      }
                  });
              });
        }
    }
}

          
```

**Comment by JKorf** [comment]
I would write the functionality like this: 
````csharp
public class BinanceCoins
{
    public ConcurrentDictionary<string, KlineDataContainer> KlineDictionary { get; }
    public ConcurrentDictionary<string, KlineDataContainer> TickerDictionary { get; }
    
    private BinanceSocketClient _socketClient;
    private List<string> _symbols;

    public BinanceCoins(List<string> symbols)
    {
        _socketClient = new BinanceSocketClient();
        _symbols = symbols;

        KlineDictionary = new ConcurrentDictionary<string, KlineDataContainer>();
        TickerDictionary = new ConcurrentDictionary<string, KlineDataContainer>();
    }

    public async Task<bool> StartAsync()
    {
        var klineSubscription = await _socketClient.SpotStreams.SubscribeToKlineUpdatesAsync(_symbols, KlineInterval.OneMinute, ProcessKlineUpdate);
        if(!klineSubscription.Success)
        {
            // Do something when subscribing is not succesful, for example
            Console.WriteLine("Failed to subscribe klines: " + klineSubscription.Error);
            return false;
        }

        var tradeSubscription = await _socketClient.SpotStreams.SubscribeToAggregatedTradeUpdatesAsync(_symbols, ProcessTradeUpdate);
        if (!tradeSubscription.Success)
        {
            // Do something when subscribing is not succesful, for example
            Console.WriteLine("Failed to subscribe trades: " + tradeSubscription.Error);

            // Close/cleanup the kline subscription since we didn't subscribe successfully
            await klineSubscription.Data.CloseAsync();
            return false;
        }

        return true;
    }

    private void ProcessKlineUpdate(DataEvent<IBinanceStreamKlineData> dataEvent)
    {
        if (KlineDictionary.TryGetValue(dataEvent.Data.Symbol, out var existingItem))
        {
            existingItem.OpenPrice = dataEvent.Data.Data.OpenPrice;
            existingItem.HighPrice = dataEvent.Data.Data.HighPrice;
            existingItem.LowPrice = dataEvent.Data.Data.LowPrice;
        }
        else
        {
            KlineDictionary.TryAdd(dataEvent.Data.Symbol, new KlineDataContainer
            {
                OpenPrice = dataEvent.Data.Data.OpenPrice,
                HighPrice = dataEvent.Data.Data.HighPrice,
                LowPrice = dataEvent.Data.Data.LowPrice,
            });
        }
    }

    private void ProcessTradeUpdate(DataEvent<BinanceStreamAggregatedTrade> dataEvent)
    {
        // Same principle as other subscription here
    }
}

public class KlineDataContainer
{
    public decimal HighPrice { get; set; }
    public decimal OpenPrice { get; set; }
    public decimal LowPrice { get; set; }

    // Any info you want to store, probably want a similar class for the trade updates
}
````

This is using the latest version of the library btw, which has been slightly changed from the version you use.

**Comment by UtkuX** [comment]
Thank you for the response and help. I was using your version 7.2.7 because it was causing incompatibility with other libraries, would that be a problem?

**Comment by JKorf** [comment]
Could be yes. If you're using other libraries from third parties which don't support the CryptoExchange.Net version 5 then you'll have to keep using 7.2.7 for now. If you're only using packages released by me then you should be able to use the latest version, but you'll have to update all those packages to the latest at the same time.

**Comment by JKorf** [comment]
If you installed the latest version while still having your 3rd party dependency then you'll get runtime errors unfortunately

**Comment by UtkuX** [comment]
> If you installed the latest version while still having your 3rd party dependency then you'll get runtime errors unfortunately

Well, without installing the latest version of Binance.Net wrapper, there is no problem in compliering the class you sent me, but will there be any problems in runtime?

**Comment by JKorf** [comment]
That doesn't make sense, in the version you posted here the streams are accessed by `socket.Spot.Subscribe`, while I use `socket.SpotStreams.Subscribe` as in the latest version. It's probably your IDE just not giving errors until you build?

**Comment by UtkuX** [comment]
yes, I've had the weakness of not using visual studio, I'm sorry. What exactly is the difference between Spot and SpotStreams, so using spot won't solve the problem I mentioned at the beginning of the topic? thanks...

**Comment by JKorf** [comment]
It's only a naming change, nothing functionally different between the two. For your initial issue, I don't immediately see an issue that would cause the warning you posted. There is no reason the processing should take 18seconds in the code you posted. So the issue is maybe in other part. Note that when you put a breakpoint in the data handler it will also give this error as it counts the time in the breakpoint as processing time.

**Comment by UtkuX** [comment]
thank you very much.

**Comment by UtkuX** [comment]
Hello, I'm disturbing you again, but let me ask this. I want to get the last 30 candles in 1 minute via rest api just before I subscribe to klineupdatedata from socket. If I had continued with Task, I would have done it with continuewith inside the constructor, but I guess I can't do that in the clean and functional code you posted. What is your suggested solution? While comparing the data of 2 or 3 other exchanges, would it be more correct to put the data from each exchange in a separate thread in program.cs? Excuse my inexperience. Thank you for your help.

**Comment by eros1453** [comment]
var klines = await bfc.UsdFuturesApi.ExchangeData.GetKlinesAsync(ASSET, KlineInterval.OneMinute, null, null, 50);
                    IEnumerable<IBinanceKline> klones = klines.Data.OrderByDescending(a => a.CloseTime).ToArray();

**Comment by UtkuX** [comment]
> OrderByDescending



> var klines = await bfc.UsdFuturesApi.ExchangeData.GetKlinesAsync(ASSET, KlineInterval.OneMinute, null, null, 50); IEnumerable klones = klines.Data.OrderByDescending(a => a.CloseTime).ToArray();

no, it won't work.

**Comment by eros1453** [comment]
Works for me

**Comment by eros1453** [comment]
Gettings klines first and start stream after.
calculating MA and other indicator with no problems.

**Comment by UtkuX** [comment]
> Gettings klines first and start stream after.
> calculating MA and other indicator with no problems.

did you use threads or tasks?

**Comment by eros1453** [comment]
Updates with Task and timer gives me a cleaner view and better timing.
Else you can have a Tradingview like seconds chart witch is useful for Furues trades where every second counts.

**Comment by UtkuX** [comment]
> Updates with Task and timer gives me a cleaner view and better timing. Else you can have a Tradingview like seconds chart witch is useful for Furues trades where every second counts. Also with Taskfactory and multiple timers you can have multiple timeframes for TI calculations and Candle Bars

Hello again, I wrote to you on Twitter because I do not trust my English and I think that the questions I asked spoil the integrity of the subject, but I did not get a response. I understand that you get second candles like in tradingview premium, which is a great thing. Can you share sample code about it? If I follow each coin with a separate object, how can I run them in parallel? that is, if I set the consturctor method as a string, not a list.
